include/utilities.h

```diff
@@ -97,57 +99,138 @@ void smart_delete (SmartPointer<TYPE> &sp)
-        }
-      else
-        {
-          Sdouble ildv = global_vector (local_dof_indices[i]);
-          ildv.diff (i, dofs_per_cell);
-          ((Sdouble &)independent_local_dofs[i]) = ildv;
-          if (typeid(Number) == typeid(SSdouble))
-            {
-              ((SSdouble &)independent_local_dofs[i]).diff(i,dofs_per_cell);
-            }
-        }
-    }
+  /**
+   *  Extract local dofs values and initialize the number
+   *  of independent variables up to the second order derivative.
+   *
+   */
+  template <typename Number, typename VEC>
+  void
+  extract_local_dofs (const VEC &global_vector,
+                      const std::vector<types::global_dof_index> &local_dof_indices,
+                      std::vector<Number> &independent_local_dofs)
+  {
+    const unsigned int dofs_per_cell = local_dof_indices.size();
+    for (unsigned int i=0; i < dofs_per_cell; ++i)
+      {
+        if (typeid(Number) == typeid(double))
+          {
+            independent_local_dofs[i] = global_vector (local_dof_indices[i]);
+          }
+        else
+          {
+            Sdouble ildv = global_vector (local_dof_indices[i]);
+            ildv.diff (i, dofs_per_cell);
+            ((Sdouble &)independent_local_dofs[i]) = ildv;
+            if (typeid(Number) == typeid(SSdouble))
+              {
+                ((SSdouble &)independent_local_dofs[i]).diff(i,dofs_per_cell);
+              }
+          }
+      }
+  }
+
+  /**
+   *  Extract independent local dofs values of a cell and store them in
+   *  std::vector <Tensor <1, dim, Number> > us
+   *  whose size is number of quadrature points in the cell.
+   *
+   */
+  template <int dim, int spacedim, typename Number>
+  void
+  get_u_values (const FEValues<dim, spacedim> &fe_values,
+                const std::vector<Number> &independent_local_dof_values,
+                std::vector <Tensor <1, dim, Number> > &us)
+  {
+    const unsigned int           dofs_per_cell = fe_values.dofs_per_cell;
+    const unsigned int           n_q_points    = fe_values.n_quadrature_points;
+
+    AssertDimension(us.size(), n_q_points);
+
+    for (unsigned int q=0; q<n_q_points; ++q)
+      {
+        for (unsigned int i=0; i<dofs_per_cell; ++i)
+          {
+            const unsigned int c = fe_values.get_fe().system_to_component_index(i).first;
+            us[q][c] += independent_local_dof_values[i]*fe_values.shape_value_component(i,q,c);
+          }
+      }
+  }
+
+  /**
+   *  Extract independent local dofs values of a cell,
+   *  compute the deformation gradient F in each quadrature point,
+   *  which is stored in
+   *  std::vector <Tensor <2, dim, Number> > Fs
+   *  whose size is number of quadrature points in the cell.
+   *
+   */
+  template <int dim, int spacedim, typename Number>
+  void
+  get_F_values (const FEValues<dim, spacedim> &fe_values,
+                const std::vector<Number> &independent_local_dof_values,
+                std::vector <Tensor <2, dim, Number> > &Fs)
```

Comment:
Perché questo sia generale, dovresti mettere una ComponentMask che indichi quali sono le componenti di cui vuoi calcolare le F, ma lo possiamo fare anche in un future PR.